docs/design/hooks.md

```diff
@@ -0,0 +1,262 @@
+update said best practices, we get it for hooks as well. It also unifies the
+user experience to simplify the setup process.
+
+### Pre-upload hooks
+
+Pre-upload hooks fill a large variety of purposes.
+
+The most common kinds of general-purpose hooks, which we will definitely need to
+support, are:
+* Validate something about the description 
+  * Only needs the commit description, no checkout required
+* Validate something about the code (eg. Run lints, or run a specific presubmit
+  tool)
+  * *May* need a checkout
+* Wants to use `jj run` under the hood
+  * Until `jj run` is implemented, the script probably needs to run
+    `jj edit <commit>`, then once it's done `jj edit <original commit>`
+  * Ideally wants a VFS
+  * Due to performance reasons, some repos may want to only run on the heads
+* Modify the description (eg. add a git footer)
+  * Doesn’t need a checkout
+  * Can rewrite commits
+
+To solve this, I propose we take a similar approach to our fix tools. Namely,
+our configuration will look like:
+
+```toml
+# Fix and sign specifically do not support commands.
+[hooks.pre-upload.fix]
+enabled = true
+order = 0 # Default (fix always runs first by default)
+[hooks.pre-upload.sign]
+enabled = true
+order = 99999 # Default (sign always runs last by default)
+[hooks.pre-upload.my-tool]
+# This defaults to enabled, similar to fix.tools.*
+command = ["python3", "$root/my_hook.py"]
+order = 1 # Default
+```
+
+Hooks will be ran in ascending order, with multiple hooks of the same order
+being an error.
+
+A hook will need to meet a spec. Specifically, `jj` will pass some json to
+`$COMMAND` via stdin:
+```json
+{
+    // Probably not required, but no harm providing this.
+    "operation": "<operation ID>"
+    // commit ID of @
+    "working_copy": "<commit ID>",
+    // All commits that we are going to upload. // In topological order.
+    "to_upload": ["<commit ID>", "<commit ID>"],
+
+    // @ | to_upload
+    "commits": {
+        "<commit ID>": {
+            // The colored format used in jj log.
+            "pretty": "...",
+            "change_id": "...",
+            "description": "...",
+            "empty": false,
+            // May contain commit IDs not in "commits".
+            "parents": ["<commit ID>", "<commit ID>"],
+        }
+    }
+}
+```
+
+The stderr of the command will be passed through to the stderr of the main
+program, and a nonzero exit code will result in the main program complaining
+that a hook failed. The command must not write to stdout, as we reserve it in
+case we want to use that mechanism to write info back to jj in a future version
+of hooks.
+
+In the initial version, we will require that hooks are (with the exception of
+the builtin hooks fix & sign), readonly.
+
+In future versions, there are a few things we could do to allow more interesting
+hooks that can write as well as read. It's unclear what the right approach would
+be:
```

Comment:
another future use-case is to run the "hooks" in the daemon process.

Reply:
What would be the advantage of such an approach? AFAICT the advantages of running something in the daemon process is that you don't have to wait for it to finish, while the disadvantages are an increased complexity of IPC, issues with signaling (eg. control-C), what to do if the starts a process that infinite loops, etc.

By a hook's very nature, you need to wait on it to finish, so I don't see the reason to do so? Am I missing some advantages?